proposals/module-types/Explainer.md

```diff
@@ -0,0 +1,745 @@
+    call $i.$out
+  )
+)
+```
+the outer module imports a module `$M` and a function `$f` and then uses `$f` to
+instantiate `$m` producing an instance `$i`. `instance` definitions are created
+using [constant initializer expression], similar to `global` definitions. The
+`instance.instantiate` instruction introduced by this proposal is *only* allowed
+in constant initializer expressions, but could in the future be allowed as a
+normal instruction to enable runtime instantiation. Its signature is:
+```
+instance.instantiate $module : [ (ref Tᵢ)ⁿ ] -> [ (ref $InstanceT) ]
+```
+where
+* `$module` refers to a module (in the module index space) of type `$ModuleT`
+* the `n` operands `Tᵢ` match the `n` imports of `$ModuleT` in import vector order
+* `$InstanceT` is the instance type of `$ModuleT`
+
+Note that while `instance.instantiate` relies on the positional order of imports
+in `$ModuleT`, this order is with respect to the *local* type definition
+`$ModuleT`. At instantiation-time, module and instance subtyping is based solely
+on strings and thus reordering, adding exports or dropping imports will never
+break clients.
+
+Instance imports and definitions can also be used as operands to
+`instance.instantiate`. For example, this module imports a `wasi_file`
+instance and passes it on to its child:
+```wasm
+(module
+  (type $WasiFile (instance
+    (export "read" (func (param i32 i32 i32) (result i32)))
+    (export "write" (func (param i32 i32 i32) (result i32)))
+  ))
+  (import "wasi_file" (instance $wasi-file (type $WasiFile)))
+  (import "child" (module $child
+    (import "wasi_file" (instance (type $WasiFile)))
+  ))
+  (instance (instance.instantiate $child (ref.instance $wasi-file)))
+)
+```
+Here, `ref.instance` is a new instruction symmetric to [`ref.func`], selecting an
+instance from the instance index space via a static index immediate. It will be
+discussed more in the next section.
+
+`instance.instantiate` allows its operands to be subtypes of the declared
+module import types. When the operands are of instance or module type, that
+means checking module/instance subtyping. Unlike with [GC types], where
+subtyping needs to imply memory layout, module and instance subtyping can be
+order-independent, matching up by import/export names instead.
+
+`instance.instantiate` can refer back to all instance imports and all preceding
+instance definitions, allowing the creation of arbitrary acyclic graphs of
+instances. The nested instances of a module `M` are instantiated during
+[`module_instantiate`]`(M)`, before the top-level `M`'s own [`moduleinst`] is
+created. Thus, `instance.instantiate` cannot refer to any functions, memories,
+tables or global defined by the top-level `M`, since they do not yet exist when
+creating nested instances. In general, with this proposal,
+`module_instantiate(M)` goes from always creating a single instance to
+potentially creating a *DAG* of instances, with `M`'s instance as the root. 
```

Comment:
Related to the previous point, a frequent situation when plugging together multiple modules from different sources is that the import of one module, even when expressed as an instance, has to combine individual values from multiple sources. It may also have to adapt names. In ML, I can write a simple adaptor structure in place to rename and regroup things, but here that would again require going through an auxiliary wrapper module that needs to be instantiated separately.

Perhaps there should be (1) a way to reference instance exports and (2) a way to create instance objects directly?

Reply:
Yeah, I've been wondering about this exact same thing, b/c yeah a way to create instances directly does make sense.  Basically, I just wonder if it's worth it: my current understanding is that small nested modules that simply reexport things with different names will be effectively free (e.g., not exist after compile time).  Given that, there's only a slight size/verbosity win, I think?

Reply:
Oh, and to address (1): I think one can sufficiently reference an instance's exports via the containing index spaces they are injected into (with `$foo.$bar`), or do you mean something else?

Reply:
Re (1): Oh, right, I somehow overlooked that that works for other things than just types or modules. But see my comment above.

Re (2): Maybe. It still seems a bit awkward. And computationally, being forced to insert a beta expansion is rather odd. Perhaps this can be deferred, though.

Reply:
Ok, it could make sense later, but let's build up that use case and defer for now.